gui/wxpython/main_window/page.py

```diff
@@ -0,0 +1,109 @@
+        self._dockable = dockable
+
+        # distinguishes whether map panel is docked or not
+        self._docked = True
+
+        # undock/dock bound method
+        self._docking_callback = None
+
+        # Emitted when switching map notebook tabs (Single-Window)
+        self.onFocus = Signal("MainPage.onFocus")
+
+        # Emitted when closing page by closing its window.
+        self.closingPage = Signal("MainPage.closingPage")
+
+        # Emitted when renaming page.
+        self.renamingPage = Signal("MainPage.renamingPage")
+
+    def _pgnumDict(self):
+        """Get dictionary containg page index"""
+        return {"mainnotebook": self._mainnotebook.GetPageIndex(self)}
+
+    def SetUpPage(self, parent, notebook, can_close=None):
+        self._mainnotebook = notebook
+
+        def CanClosePage():
+            return self._pgnumDict()
+
+        # set callbacks
+        self.canCloseCallback = CanClosePage if can_close is None else can_close
+        self.SetDockingCallback(notebook.UndockPage)
+
+        # bind various events
+        self.closingPage.connect(parent._closePageNoEvent)
+        self.renamingPage.connect(parent._renamePageNoEvent)
+
+    def SetDockingCallback(self, function):
+        """Sets docking bound method to dock or undock"""
+        self._docking_callback = function
+
+    def IsDocked(self):
+        return self._docked
+
+    def IsDockable(self):
+        return self._dockable
+
+    def OnDockUndock(self, event=None):
+        """Dock or undock map display panel to independent MapFrame"""
+        if self._docking_callback:
+            self._docked = not self._docked
+            self._docking_callback(self)
+
+    def _onCloseWindow(self, event):
+        """Close window"""
+        if self.canCloseCallback:
+            pgnum_dict = self.canCloseCallback()
+            if pgnum_dict is not None:
+                if self.IsDockable():
+                    self.closingPage.emit(
+                        pgnum_dict=pgnum_dict, is_docked=self.IsDocked()
+                    )
+                    if not self.IsDocked():
+                        frame = self.GetParent()  # pylint: disable=E1101
+                        frame.Destroy()
+                else:
+                    self.closingPage.emit(pgnum_dict=pgnum_dict)
+                # Destroy is called when notebook page is deleted
+        else:
+            self.parent.Destroy()  # pylint: disable=E1101
+
+    def RenamePage(self, title):
+        """Rename page or change frame title"""
+        if self.canCloseCallback:
+            pgnum_dict = self._pgnumDict()
+            if pgnum_dict is not None:
+                if self.IsDockable():
+                    self.renamingPage.emit(
+                        pgnum_dict=pgnum_dict, is_docked=self.IsDocked(), text=title
+                    )
+                    if not self.IsDocked():
+                        self.GetParent().SetTitle(title)  # pylint: disable=E1101
```

Comment:
I think the disable is justified in this case, but the inline solution does not look right given there is a couple of those. Hard to say what is the correct solution for disabling here. Perhaps add that class to:

```
gui/wxpython/.pylintrc:322:ignored-classes=optparse.Values,thread._local,_thread._local
```

http://pylint-messages.wikidot.com/messages:e1101
https://pylint.pycqa.org/en/latest/user_guide/configuration/all-options.html#ignored-classes

Reply:
Changed in [f49a473](https://github.com/OSGeo/grass/pull/3003/commits/f49a47379da5af03ad08467375a9f82a4d9d74aa)

Reply:
Thanks. I would say, code looks better, so let's see how this works out in the future.